File: src/S3/S3Router.php
<?php

declare(strict_types=1);

namespace MiniS3\S3;

use MiniS3\Auth\AuthException;
use MiniS3\Auth\SigV4Authenticator;
use MiniS3\Http\RequestContext;
use MiniS3\Storage\FileStorage;
use RuntimeException;
use SimpleXMLElement;
use Throwable;

final class S3Router
{
    public function __construct(
        private readonly RequestContext $request,
        private readonly FileStorage $storage,
        private readonly S3Response $response,
        private readonly RequestValidator $validator,
        private readonly SigV4Authenticator $authenticator,
        private readonly int $maxRequestSize,
        private readonly bool $publicReadAllBuckets = false
    ) {
    }

    public function handle(): never
    {
        try {
            $this->sendCorsHeaders();

            $this->storage->ensureDataDirExists();

            [$bucket, $key] = $this->extractBucketAndKey();
            $method = $this->request->getMethod();

            if ($method === 'OPTIONS') {
                http_response_code(204);
                exit;
            }

            $this->validateBucketAndKey($method, $bucket, $key);
            $this->validateRequestSize();

            if (!$this->isPublicRead($method)) {
                $this->authenticator->authenticate($this->request);
            }

            switch ($method) {
                case 'PUT':
                    $this->handlePut($bucket, $key);
                    break;

                case 'POST':
                    $this->handlePost($bucket, $key);
                    break;

                case 'GET':
                    $this->handleGet($bucket, $key);
                    break;

                case 'HEAD':
                    $this->handleHead($bucket, $key);
                    break;

                case 'DELETE':
                    $this->handleDelete($bucket, $key);
                    break;

                default:
                    $this->response->error(405, 'MethodNotAllowed', 'Method not allowed', $this->resource($bucket, $key));
            }
        } catch (AuthException $e) {
            $this->response->error($e->getHttpStatus(), $e->getS3Code(), $e->getMessage());
        } catch (Throwable $e) {
            $this->response->error(500, 'InternalError', 'Internal server error');
        }
    }

    private function isPublicRead(string $method): bool
    {
        return $this->publicReadAllBuckets && ($method === 'GET' || $method === 'HEAD');
    }

    private function sendCorsHeaders(): void
    {
        $origin = $this->request->getHeader('origin');
        if ($origin === null || $origin === '') {
            return;
        }

        header('Access-Control-Allow-Origin: ' . $origin);
        header('Vary: Origin');
        header('Access-Control-Allow-Methods: GET, HEAD, PUT, POST, DELETE, OPTIONS');
        header('Access-Control-Expose-Headers: ETag, Content-Length, Content-Range');

        $requestHeaders = $this->request->getHeader('access-control-request-headers');
        if ($requestHeaders !== null && $requestHeaders !== '') {
            header('Access-Control-Allow-Headers: ' . $requestHeaders);
        }

        header('Access-Control-Max-Age: 86400');
    }

    private function handlePut(string $bucket, string $key): never
    {
        $uploadId = $this->request->getQueryParam('uploadId');
        $partNumber = $this->request->getQueryParam('partNumber');

        if ($uploadId !== null && $partNumber !== null) {
            if (!$this->validator->isPositiveInteger($partNumber)) {
                $this->response->error(400, 'InvalidPart', 'partNumber must be a positive integer', $this->resource($bucket, $key));
            }

            if (!$this->storage->multipartDirExists($bucket, $key, $uploadId)) {
                $this->response->error(404, 'NoSuchUpload', 'Upload ID not found', $this->resource($bucket, $key));
            }

            $partPath = $this->storage->putMultipartPartFromInput($bucket, $key, $uploadId, (int) $partNumber);
            header('ETag: ' . md5_file($partPath));
            http_response_code(200);
            exit;
        }

        $this->storage->putObjectFromInput($bucket, $key);
        http_response_code(200);
        exit;
    }

    private function handlePost(string $bucket, string $key): never
    {
        if ($this->request->hasQueryParam('delete')) {
            $xml = $this->parseXmlBody();

            $quiet = isset($xml->Quiet) && strtolower((string) $xml->Quiet) === 'true';
            $deletedKeys = [];
            $errors = [];

            foreach ($xml->Object as $object) {
                $objectKey = (string) $object->Key;
                if (!$this->validator->isValidObjectKey($objectKey)) {
                    $errors[] = [
                        'key' => $objectKey,
                        'code' => 'InvalidObjectKey',
                        'message' => 'Invalid object key',
                    ];
                    continue;
                }

                $this->storage->deleteObject($bucket, $objectKey);
                if (!$quiet) {
                    $deletedKeys[] = $objectKey;
                }
            }

            $this->response->deleteResult($deletedKeys, $errors);
        }

        if ($this->request->hasQueryParam('uploads')) {
            $uploadId = bin2hex(random_bytes(16));
            $this->storage->createMultipartUpload($bucket, $key, $uploadId);
            $this->response->createMultipartUpload($bucket, $key, $uploadId);
        }

        $uploadId = $this->request->getQueryParam('uploadId');
        if ($uploadId !== null) {
            if (!$this->storage->multipartDirExists($bucket, $key, $uploadId)) {
                $this->response->error(404, 'NoSuchUpload', 'Upload ID not found', $this->resource($bucket, $key));
            }

            $xml = $this->parseXmlBody();
            $parts = [];
            foreach ($xml->Part as $part) {
                $partNumber = (string) $part->PartNumber;
                if (!$this->validator->isPositiveInteger($partNumber)) {
                    $this->response->error(400, 'InvalidPart', 'PartNumber must be a positive integer', $this->resource($bucket, $key));
                }

                $parts[] = (int) $partNumber;
            }

            if ($parts === []) {
                $this->response->error(400, 'InvalidPart', 'Multipart completion request has no parts', $this->resource($bucket, $key));
            }

            $parts = array_values(array_unique($parts));
            sort($parts, SORT_NUMERIC);

            try {
                $this->storage->completeMultipartUpload($bucket, $key, $uploadId, $parts);
                $this->storage->cleanupMultipartUpload($bucket, $key, $uploadId);
            } catch (RuntimeException $e) {
                if (str_starts_with($e->getMessage(), 'Part file missing')) {
                    $this->response->error(400, 'InvalidPart', $e->getMessage(), $this->resource($bucket, $key));
                }
                throw $e;
            }

            $scheme = (!empty($_SERVER['HTTPS']) && $_SERVER['HTTPS'] !== 'off') ? 'https' : 'http';
            $this->response->completeMultipartUpload($bucket, $key, $uploadId, $this->request->getHost(), $scheme);
        }

        $this->response->error(
            400,
            'InvalidRequest',
            'Invalid POST request: missing delete, uploads or uploadId parameter',
            $this->resource($bucket, $key)
        );
    }

    private function handleGet(string $bucket, string $key): never
    {
        if ($key === '') {
            $prefix = (string) ($this->request->getQueryParam('prefix') ?? '');
            $files = $this->storage->listFiles($bucket, $prefix);
            $this->response->listObjects($files, $bucket, $prefix);
        }

        $metadata = $this->storage->objectMetadata($bucket, $key);
        if ($metadata === null) {
            $this->response->error(404, 'NoSuchKey', 'Object not found', $this->resource($bucket, $key));
        }

        $fileSize = (int) $metadata['size'];
        $fp = $this->storage->openObjectReadStream($bucket, $key);
        $mimeType = (string) $metadata['mimeType'];
        $range = $this->request->getHeader('range');

        $start = 0;
        $end = max(0, $fileSize - 1);
        $length = $fileSize;
        $status = 200;

        if ($range !== null && $range !== '') {
            [$isValidRange, $start, $end] = $this->validator->parseRange($range, $fileSize);
            if (!$isValidRange) {
                fclose($fp);
                $this->response->sendInvalidRangeHeader($fileSize);
                http_response_code(416);
                exit;
            }

            $status = 206;
            $length = ($fileSize === 0) ? 0 : ($end - $start + 1);
            $this->response->sendRangeHeader($start, $end, $fileSize);
        }

        $this->response->sendObjectHeaders($status, $length, $mimeType, basename($key));

        if ($length === 0) {
            fclose($fp);
            exit;
        }

        fseek($fp, $start);

        $remaining = $length;
        $chunkSize = 8 * 1024 * 1024;

        while (!feof($fp) && $remaining > 0 && (PHP_SAPI === 'cli' || connection_aborted() === 0)) {
            $buffer = fread($fp, min($chunkSize, $remaining));
            if ($buffer === false) {
                break;
            }

            echo $buffer;
            $remaining -= strlen($buffer);
            flush();
        }

        fclose($fp);
        exit;
    }

    private function handleHead(string $bucket, string $key): never
    {
        if ($key === '') {
            $this->response->error(400, 'InvalidRequest', 'Object key required for HEAD', $this->resource($bucket, $key));
        }

        $metadata = $this->storage->objectMetadata($bucket, $key);
        if ($metadata === null) {
            $this->response->error(404, 'NoSuchKey', 'Resource not found', $this->resource($bucket, $key));
        }

        $this->response->sendObjectHeaders(200, (int) $metadata['size'], (string) $metadata['mimeType'], basename($key), false);
        exit;
    }

    private function handleDelete(string $bucket, string $key): never
    {
        $uploadId = $this->request->getQueryParam('uploadId');
        if ($uploadId !== null) {
            if (!$this->storage->multipartDirExists($bucket, $key, $uploadId)) {
                $this->response->error(404, 'NoSuchUpload', 'Upload ID not found', $this->resource($bucket, $key));
            }

            $this->storage->abortMultipartUpload($bucket, $key, $uploadId);
            http_response_code(204);
            exit;
        }

        $this->storage->deleteObject($bucket, $key);
        http_response_code(204);
        exit;
    }

    private function validateBucketAndKey(string $method, string $bucket, string $key): void
    {
        if ($method !== 'GET' && $bucket === '') {
            $this->response->error(400, 'InvalidBucketName', 'Bucket name not specified', '/');
        }

        if ($method === 'GET' && $bucket === '') {
            $this->response->error(400, 'InvalidBucketName', 'Bucket name required', '/');
        }

        if ($bucket !== '' && !$this->validator->isValidBucketName($bucket)) {
            $this->response->error(400, 'InvalidBucketName', 'Invalid bucket name', '/' . $bucket);
        }

        if ($key !== '' && !$this->validator->isValidObjectKey($key)) {
            $this->response->error(400, 'InvalidObjectKey', 'Invalid object key', $this->resource($bucket, $key));
        }
    }

    private function validateRequestSize(): void
    {
        if ($this->validator->isOversizedRequest($this->request->getHeader('content-length'), $this->maxRequestSize)) {
            $this->response->error(413, 'EntityTooLarge', 'Request too large');
        }
    }

    private function extractBucketAndKey(): array
    {
        $trimmedPath = trim($this->request->getPath(), '/');
        if ($trimmedPath === '') {
            return ['', ''];
        }

        $rawParts = explode('/', $trimmedPath);
        $decodedParts = array_map('rawurldecode', $rawParts);

        $bucket = $decodedParts[0] ?? '';
        $key = implode('/', array_slice($decodedParts, 1));

        return [$bucket, $key];
    }

    private function parseXmlBody(): SimpleXMLElement
    {
        $inputStream = PHP_SAPI === 'cli' ? 'php://stdin' : 'php://input';
        $body = file_get_contents($inputStream);
        if (!is_string($body) || trim($body) === '') {
            $this->response->error(400, 'MalformedXML', 'The XML you provided was not well-formed or did not validate against our published schema.');
        }

        $previous = libxml_use_internal_errors(true);
        $xml = simplexml_load_string($body);
        libxml_use_internal_errors($previous);

        if ($xml === false) {
            $this->response->error(400, 'MalformedXML', 'The XML you provided was not well-formed or did not validate against our published schema.');
        }

        return $xml;
    }

    private function resource(string $bucket, string $key): string
    {
        if ($bucket === '') {
            return '/';
        }

        if ($key === '') {
            return '/' . $bucket;
        }

        return '/' . $bucket . '/' . $key;
    }
}

File: go/internal/s3/s3_router.go
package s3

import (
	"crypto/md5"
	"crypto/rand"
	"encoding/hex"
	"encoding/xml"
	"fmt"
	"io"
	"net/http"
	"os"
	"sort"
	"strings"

	"github.com/fadlee/mini-s3/internal/auth"
	minihttp "github.com/fadlee/mini-s3/internal/http"
	"github.com/fadlee/mini-s3/internal/storage"
)

// S3Router dispatches S3-compatible HTTP requests, mirroring
// MiniS3\S3\S3Router from the PHP reference.
type S3Router struct {
	storage             *storage.FileStorage
	validator           *RequestValidator
	authenticator       *auth.SigV4Authenticator
	maxRequestSize      int64
	publicReadAllBuckets bool
}

// New creates an S3Router.
func New(st *storage.FileStorage, authenticator *auth.SigV4Authenticator, maxRequestSize int64, publicReadAllBuckets bool) *S3Router {
	return &S3Router{
		storage:             st,
		validator:           &RequestValidator{},
		authenticator:       authenticator,
		maxRequestSize:      maxRequestSize,
		publicReadAllBuckets: publicReadAllBuckets,
	}
}

// ServeHTTP implements http.Handler.
func (r *S3Router) ServeHTTP(w http.ResponseWriter, req *http.Request) {
	ctx := minihttp.NewRequestContext(req)
	resp := NewS3Response(w)

	r.sendCorsHeaders(w, ctx)

	if err := r.storage.EnsureDataDirExists(); err != nil {
		resp.XMLError(500, "InternalError", "Internal server error", "")
		return
	}

	bucket, key := r.extractBucketAndKey(ctx)
	method := ctx.Method()

	if method == "OPTIONS" {
		w.WriteHeader(http.StatusNoContent)
		return
	}

	if err := r.validateBucketAndKey(method, bucket, key, resp); err != nil {
		return
	}

	if r.isOversizedRequest(ctx) {
		resp.XMLError(413, "EntityTooLarge", "Request too large", "")
		return
	}

	if !r.isPublicRead(method) {
		if err := r.authenticator.Authenticate(ctx); err != nil {
			if ae, ok := err.(*auth.Exception); ok {
				resp.XMLError(ae.HTTPStatus, ae.S3Code, ae.Message, "")
			} else {
				resp.XMLError(500, "InternalError", "Internal server error", "")
			}
			return
		}
	}

	switch method {
	case "PUT":
		r.handlePut(ctx, resp, bucket, key)
	case "POST":
		r.handlePost(ctx, resp, bucket, key)
	case "GET":
		r.handleGet(ctx, resp, bucket, key)
	case "HEAD":
		r.handleHead(ctx, resp, bucket, key)
	case "DELETE":
		r.handleDelete(ctx, resp, bucket, key)
	default:
		resp.XMLError(405, "MethodNotAllowed", "Method not allowed", r.resource(bucket, key))
	}
}

func (r *S3Router) isPublicRead(method string) bool {
	return r.publicReadAllBuckets && (method == "GET" || method == "HEAD")
}

func (r *S3Router) sendCorsHeaders(w http.ResponseWriter, ctx *minihttp.RequestContext) {
	origin := ctx.GetHeader("origin")
	if origin == "" {
		return
	}
	w.Header().Set("Access-Control-Allow-Origin", origin)
	w.Header().Add("Vary", "Origin")
	w.Header().Set("Access-Control-Allow-Methods", "GET, HEAD, PUT, POST, DELETE, OPTIONS")
	w.Header().Set("Access-Control-Expose-Headers", "ETag, Content-Length, Content-Range")

	requestHeaders := ctx.GetHeader("access-control-request-headers")
	if requestHeaders != "" {
		w.Header().Set("Access-Control-Allow-Headers", requestHeaders)
	}
	w.Header().Set("Access-Control-Max-Age", "86400")
}

func (r *S3Router) handlePut(ctx *minihttp.RequestContext, resp *S3Response, bucket, key string) {
	uploadID := ctx.GetQueryParam("uploadId")
	partNumber := ctx.GetQueryParam("partNumber")

	if uploadID != "" && partNumber != "" {
		if !r.validator.IsPositiveInteger(partNumber) {
			resp.XMLError(400, "InvalidPart", "partNumber must be a positive integer", r.resource(bucket, key))
			return
		}
		if !r.storage.MultipartDirExists(bucket, key, uploadID) {
			resp.XMLError(404, "NoSuchUpload", "Upload ID not found", r.resource(bucket, key))
			return
		}

		partNum := atoiSafe(partNumber)
		partPath, err := r.storage.PutMultipartPart(bucket, key, uploadID, partNum, ctx.Request().Body)
		if err != nil {
			resp.XMLError(500, "InternalError", "Internal server error", r.resource(bucket, key))
			return
		}

		etag := md5File(partPath)
		resp.Writer().Header().Set("ETag", etag)
		resp.Writer().WriteHeader(http.StatusOK)
		return
	}

	if err := r.storage.PutObject(bucket, key, ctx.Request().Body); err != nil {
		resp.XMLError(500, "InternalError", "Internal server error", r.resource(bucket, key))
		return
	}
	resp.Writer().WriteHeader(http.StatusOK)
}

func (r *S3Router) handlePost(ctx *minihttp.RequestContext, resp *S3Response, bucket, key string) {
	if ctx.HasQueryParam("delete") {
		r.handleBulkDelete(ctx, resp, bucket, key)
		return
	}

	if ctx.HasQueryParam("uploads") {
		uploadID := randomHex(16)
		if err := r.storage.CreateMultipartUpload(bucket, key, uploadID); err != nil {
			resp.XMLError(500, "InternalError", "Internal server error", r.resource(bucket, key))
			return
		}
		resp.CreateMultipartUpload(bucket, key, uploadID)
		return
	}

	uploadID := ctx.GetQueryParam("uploadId")
	if uploadID != "" {
		r.handleCompleteMultipart(ctx, resp, bucket, key, uploadID)
		return
	}

	resp.XMLError(400, "InvalidRequest", "Invalid POST request: missing delete, uploads or uploadId parameter", r.resource(bucket, key))
}

func (r *S3Router) handleBulkDelete(ctx *minihttp.RequestContext, resp *S3Response, bucket, key string) {
	body, err := io.ReadAll(ctx.Request().Body)
	if err != nil || strings.TrimSpace(string(body)) == "" {
		resp.XMLError(400, "MalformedXML", "The XML you provided was not well-formed or did not validate against our published schema.", r.resource(bucket, key))
		return
	}

	type object struct {
		Key string `xml:"Key"`
	}
	type deleteRequest struct {
		XMLName xml.Name `xml:"Delete"`
		Quiet   string   `xml:"Quiet"`
		Objects []object `xml:"Object"`
	}

	var req deleteRequest
	if err := xml.Unmarshal(body, &req); err != nil {
		resp.XMLError(400, "MalformedXML", "The XML you provided was not well-formed or did not validate against our published schema.", r.resource(bucket, key))
		return
	}

	quiet := strings.ToLower(req.Quiet) == "true"
	var deletedKeys []string
	var errors []DeleteError

	for _, obj := range req.Objects {
		objectKey := obj.Key
		if !r.validator.IsValidObjectKey(objectKey) {
			errors = append(errors, DeleteError{
				Key:     objectKey,
				Code:    "InvalidObjectKey",
				Message: "Invalid object key",
			})
			continue
		}
		r.storage.DeleteObject(bucket, objectKey)
		if !quiet {
			deletedKeys = append(deletedKeys, objectKey)
		}
	}

	resp.DeleteResult(deletedKeys, errors)
}

func (r *S3Router) handleCompleteMultipart(ctx *minihttp.RequestContext, resp *S3Response, bucket, key, uploadID string) {
	if !r.storage.MultipartDirExists(bucket, key, uploadID) {
		resp.XMLError(404, "NoSuchUpload", "Upload ID not found", r.resource(bucket, key))
		return
	}

	body, err := io.ReadAll(ctx.Request().Body)
	if err != nil || strings.TrimSpace(string(body)) == "" {
		resp.XMLError(400, "MalformedXML", "The XML you provided was not well-formed or did not validate against our published schema.", r.resource(bucket, key))
		return
	}

	type part struct {
		PartNumber string `xml:"PartNumber"`
	}
	type completeRequest struct {
		XMLName xml.Name `xml:"CompleteMultipartUpload"`
		Parts   []part   `xml:"Part"`
	}

	var req completeRequest
	if err := xml.Unmarshal(body, &req); err != nil {
		resp.XMLError(400, "MalformedXML", "The XML you provided was not well-formed or did not validate against our published schema.", r.resource(bucket, key))
		return
	}

	var parts []int
	for _, p := range req.Parts {
		if !r.validator.IsPositiveInteger(p.PartNumber) {
			resp.XMLError(400, "InvalidPart", "PartNumber must be a positive integer", r.resource(bucket, key))
			return
		}
		parts = append(parts, atoiSafe(p.PartNumber))
	}

	if len(parts) == 0 {
		resp.XMLError(400, "InvalidPart", "Multipart completion request has no parts", r.resource(bucket, key))
		return
	}

	parts = uniqueInts(parts)
	sort.Ints(parts)

	if err := r.storage.CompleteMultipartUpload(bucket, key, uploadID, parts); err != nil {
		if strings.HasPrefix(err.Error(), "Part file missing") {
			resp.XMLError(400, "InvalidPart", err.Error(), r.resource(bucket, key))
			return
		}
		resp.XMLError(500, "InternalError", "Internal server error", r.resource(bucket, key))
		return
	}

	r.storage.CleanupMultipartUpload(bucket, key, uploadID)

	scheme := ctx.GetScheme()
	resp.CompleteMultipartUpload(bucket, key, uploadID, ctx.GetHost(), scheme)
}

func (r *S3Router) handleGet(ctx *minihttp.RequestContext, resp *S3Response, bucket, key string) {
	if key == "" {
		prefix := ctx.GetQueryParam("prefix")
		files, err := r.storage.ListFiles(bucket, prefix)
		if err != nil {
			resp.XMLError(500, "InternalError", "Internal server error", r.resource(bucket, key))
			return
		}
		resp.ListObjects(files, bucket, prefix)
		return
	}

	metadata, err := r.storage.ObjectMetadata(bucket, key)
	if err != nil {
		resp.XMLError(500, "InternalError", "Internal server error", r.resource(bucket, key))
		return
	}
	if metadata == nil {
		resp.XMLError(404, "NoSuchKey", "Object not found", r.resource(bucket, key))
		return
	}

	fileSize := metadata.Size
	f, err := r.storage.OpenObjectReadStream(bucket, key)
	if err != nil {
		resp.XMLError(500, "InternalError", "Internal server error", r.resource(bucket, key))
		return
	}
	defer f.Close()

	mimeType := metadata.MimeType
	rangeHeader := ctx.GetHeader("range")

	start := int64(0)
	end := fileSize - 1
	if end < 0 {
		end = 0
	}
	length := fileSize
	status := 200

	if rangeHeader != "" {
		result := r.validator.ParseRange(rangeHeader, fileSize)
		if !result.Valid {
			resp.SendInvalidRangeHeader(fileSize)
			resp.Writer().WriteHeader(416)
			return
		}
		status = 206
		start = result.Start
		end = result.End
		if fileSize == 0 {
			length = 0
		} else {
			length = end - start + 1
		}
		resp.SendRangeHeader(start, end, fileSize)
	}

	resp.SendObjectHeaders(status, length, mimeType, baseName(key), true)

	if length == 0 {
		return
	}

	f.Seek(start, io.SeekStart)
	io.CopyN(resp.Writer(), f, length)
}

func (r *S3Router) handleHead(ctx *minihttp.RequestContext, resp *S3Response, bucket, key string) {
	if key == "" {
		resp.XMLError(400, "InvalidRequest", "Object key required for HEAD", r.resource(bucket, key))
		return
	}

	metadata, err := r.storage.ObjectMetadata(bucket, key)
	if err != nil {
		resp.XMLError(500, "InternalError", "Internal server error", r.resource(bucket, key))
		return
	}
	if metadata == nil {
		resp.XMLError(404, "NoSuchKey", "Resource not found", r.resource(bucket, key))
		return
	}

	resp.SendObjectHeaders(200, metadata.Size, metadata.MimeType, baseName(key), false)
}

func (r *S3Router) handleDelete(ctx *minihttp.RequestContext, resp *S3Response, bucket, key string) {
	uploadID := ctx.GetQueryParam("uploadId")
	if uploadID != "" {
		if !r.storage.MultipartDirExists(bucket, key, uploadID) {
			resp.XMLError(404, "NoSuchUpload", "Upload ID not found", r.resource(bucket, key))
			return
		}
		r.storage.AbortMultipartUpload(bucket, key, uploadID)
		resp.Writer().WriteHeader(http.StatusNoContent)
		return
	}

	r.storage.DeleteObject(bucket, key)
	resp.Writer().WriteHeader(http.StatusNoContent)
}

func (r *S3Router) validateBucketAndKey(method, bucket, key string, resp *S3Response) error {
	if method != "GET" && bucket == "" {
		resp.XMLError(400, "InvalidBucketName", "Bucket name not specified", "/")
		return fmt.Errorf("bucket name not specified")
	}
	if method == "GET" && bucket == "" {
		resp.XMLError(400, "InvalidBucketName", "Bucket name required", "/")
		return fmt.Errorf("bucket name required")
	}
	if bucket != "" && !r.validator.IsValidBucketName(bucket) {
		resp.XMLError(400, "InvalidBucketName", "Invalid bucket name", "/"+bucket)
		return fmt.Errorf("invalid bucket name")
	}
	if key != "" && !r.validator.IsValidObjectKey(key) {
		resp.XMLError(400, "InvalidObjectKey", "Invalid object key", r.resource(bucket, key))
		return fmt.Errorf("invalid object key")
	}
	return nil
}

func (r *S3Router) isOversizedRequest(ctx *minihttp.RequestContext) bool {
	contentLength := ctx.GetHeader("content-length")
	return r.validator.IsOversizedRequest(contentLength, r.maxRequestSize)
}

func (r *S3Router) extractBucketAndKey(ctx *minihttp.RequestContext) (string, string) {
	trimmedPath := strings.Trim(ctx.Path(), "/")
	if trimmedPath == "" {
		return "", ""
	}

	rawParts := strings.Split(trimmedPath, "/")
	decodedParts := make([]string, len(rawParts))
	for i, part := range rawParts {
		decodedParts[i] = decodeURIComponent(part)
	}

	bucket := decodedParts[0]
	key := strings.Join(decodedParts[1:], "/")
	return bucket, key
}

func (r *S3Router) resource(bucket, key string) string {
	if bucket == "" {
		return "/"
	}
	if key == "" {
		return "/" + bucket
	}
	return "/" + bucket + "/" + key
}

// --- helpers ---

func md5File(path string) string {
	f, err := os.Open(path)
	if err != nil {
		return ""
	}
	defer f.Close()
	h := md5.New()
	io.Copy(h, f)
	return hex.EncodeToString(h.Sum(nil))
}

func randomHex(n int) string {
	b := make([]byte, n)
	rand.Read(b)
	return hex.EncodeToString(b)
}

func baseName(key string) string {
	if idx := strings.LastIndex(key, "/"); idx != -1 {
		return key[idx+1:]
	}
	return key
}

func atoiSafe(s string) int {
	var n int
	for _, c := range s {
		if c < '0' || c > '9' {
			return 0
		}
		n = n*10 + int(c-'0')
	}
	return n
}

func uniqueInts(slice []int) []int {
	seen := make(map[int]bool)
	var result []int
	for _, v := range slice {
		if !seen[v] {
			seen[v] = true
			result = append(result, v)
		}
	}
	return result
}

func decodeURIComponent(s string) string {
	// Simple percent-decoding for path segments.
	var buf strings.Builder
	for i := 0; i < len(s); i++ {
		if s[i] == '%' && i+2 < len(s) {
			h := s[i+1 : i+3]
			b, err := hex.DecodeString(h)
			if err == nil && len(b) == 1 {
				buf.WriteByte(b[0])
				i += 2
				continue
			}
		}
		buf.WriteByte(s[i])
	}
	return buf.String()
}
